pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/scheduler/NamespaceUnloadScheduler.java

```diff
@@ -0,0 +1,155 @@
+    }
+
+    @VisibleForTesting
+    protected NamespaceUnloadScheduler(ScheduledExecutorService loadManagerExecutor,
+                                       LoadManagerContext context,
+                                       ServiceUnitStateChannel channel,
+                                       NamespaceUnloadStrategy strategy) {
+        this.namespaceUnloadStrategy = strategy;
+        this.recentlyUnloadedBundles = new ConcurrentHashMap<>();
+        this.recentlyUnloadedBrokers = new ConcurrentHashMap<>();
+        this.loadManagerExecutor = loadManagerExecutor;
+        this.context = context;
+        this.conf = context.brokerConfiguration();
+        this.channel = channel;
+    }
+
+    @Override
+    public synchronized void execute() {
+        boolean debugMode = conf.isLoadBalancerDebugModeEnabled() || log.isDebugEnabled();
+        if (debugMode) {
+            log.info("Load balancer enabled: {}, Shedding enabled: {}.",
+                    conf.isLoadBalancerEnabled(), conf.isLoadBalancerSheddingEnabled());
+        }
+        if (currentRunningFuture != null && !currentRunningFuture.isDone()
+                || !(conf.isLoadBalancerEnabled() && conf.isLoadBalancerSheddingEnabled())) {
+            return;
+        }
+        this.currentRunningFuture = channel.isChannelOwnerAsync().thenCompose(isChannelOwner -> {
+            if (!isChannelOwner) {
+                return CompletableFuture.completedFuture(null);
+            }
+            return context.brokerRegistry().getAvailableBrokersAsync().thenCompose(availableBrokers -> {
+                if (debugMode) {
+                   log.info("Available brokers: {}", availableBrokers);
+                }
+                if (availableBrokers.size() <= 1) {
+                    log.info("Only 1 broker available: no load shedding will be performed");
+                    return CompletableFuture.completedFuture(null);
+                }
+                final UnloadDecision unloadDecision = namespaceUnloadStrategy
+                        .findBundlesForUnloading(context, recentlyUnloadedBundles, recentlyUnloadedBrokers);
+                if (debugMode) {
+                    log.info("[{}] Unload decision result: {}",
+                            namespaceUnloadStrategy.getClass().getSimpleName(), unloadDecision.toString());
+                }
+                if (unloadDecision.getUnloads().isEmpty()) {
+                    return CompletableFuture.completedFuture(null);
+                }
+                List<CompletableFuture<Void>> futures = new ArrayList<>();
+                unloadDecision.getUnloads().forEach((broker, unload) -> {
+                    log.info("[{}] Unloading bundle: {}", namespaceUnloadStrategy.getClass().getSimpleName(), unload);
+                    futures.add(channel.publishUnloadEventAsync(unload).thenAccept(__ -> {
+                        recentlyUnloadedBundles.put(unload.serviceUnit(), System.currentTimeMillis());
```

Comment:
Update the recentlyUpdatedBrokers map too.

Reply:
Also, we currently don't remove items from recentlyUnloadedBundles. We need to add the removal logic(allow bundle unload again after cooldown) in shedder strategy(preferable) or in the reporter.